exp/query/api.go

```diff
@@ -88,44 +110,184 @@ func (a *API) Search(ctx context.Context, req *SearchRequest) (*SearchResponse,
 	if err != nil {
-		return nil, fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
+		return nil, nil, fmt.Errorf("failed to vectorize the nearText query: %w", err)
 	}
 
-	if err := store.CreateOrLoadBucket(ctx, defaultLSMObjectsBucket); err != nil {
-		return nil, fmt.Errorf("failed to load objects bucket in store: %w", err)
+	// TODO(kavi): Assuming BQ compression is enabled. Make it generic.
+	bq := flatent.CompressionUserConfig{
+		Enabled: true,
+	}
+
+	// NOTE(kavi): Accept distance provider as dependency?
+	dis := distancer.NewCosineDistanceProvider()
+	index, err := flat.New(flat.Config{
+		ID:               defaultLSMVectorsCompressedBucket,
+		DistanceProvider: dis,
+		RootPath:         localPath,
+	}, flatent.UserConfig{
+		BQ: bq,
+	}, store)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to initialize index: %w", err)
+	}
+	defer index.Shutdown(ctx)
+
+	// TODO(kavi): Here `limit` is not what you expect. It's the maxLimit.
+	// Currently `SearchByVectorDistance` api takes limit via `newSearchByDistParams(maxLimit)` which caller
+	// don't have control too.
+	certainty := 1 - threshold
+	matched_ids, distance, err := index.SearchByVectorDistance(vectors, certainty, int64(limit), nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to search by vector: %w", err)
+	}
+
+	opts := []lsmkv.BucketOption{
+		// lsmkv.WithStrategy(lsmkv.StrategyReplace),
+		lsmkv.WithSecondaryIndices(2),
+	}
+
+	if err := store.CreateOrLoadBucket(ctx, defaultLSMObjectsBucket, opts...); err != nil {
+		return nil, nil, fmt.Errorf("failed to vectorize query text: %w", err)
 	}
 
 	bkt := store.Bucket(defaultLSMObjectsBucket)
 
-	var resp SearchResponse
+	objs := make([]*storobj.Object, 0)
 
-	if err := bkt.IterateObjects(ctx, func(object *storobj.Object) error {
-		resp.objects = append(resp.objects, object)
-		return nil
-	}); err != nil {
-		return nil, fmt.Errorf("failed to iterate objects in store: %w", err)
+	for _, id := range matched_ids {
+		key, err := indexDocIDToLSMKey(id)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to serialize ids returned from flat index: %w", err)
+		}
+		objB, err := bkt.GetBySecondary(0, key)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get object from store: %w", err)
+		}
+
+		obj, err := storobj.FromBinary(objB)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to decode object from store: %w", err)
+		}
+
+		objs = append(objs, obj)
 	}
 
-	return &resp, nil
+	return objs, distance, nil
+}
+
+func (a *API) loadOrDownloadLSM(ctx context.Context, collection, tenant string) (*lsmkv.Store, string, error) {
+	localPath := path.Join(a.offload.DataPath, strings.ToLower(collection), strings.ToLower(tenant), defaultLSMRoot)
+
+	// NOTE: Download only if path doesn't exist.
+	// Assumes, whatever in the path is latest.
+	// We will add a another way to keep this path upto date via some data versioning.
+	a.omu.Lock()
```

Comment:
Maybe we can make it a sync map to lock, when the key in the sync map is the class+tenant string. This way we only lock for the same class/tenant request and don't block request for any other class/tenant.

What do you think ?

Reply:
> This way we only lock for the same class/tenant request and don't block request for any other class/tenant.
that's smart. will do 👍 

My small concern here is the memory footprint of those map. Entry for `class+tenant` stays in memory forever even though tenant is less engaged (meaning we just needed to download once). Unless we introduce some mechanism to cleanup the map once in a while.

Reply:
I agree with you that a good concern. What we can do is the following
1. Update the metrics ticket to track the amount of class+tenant we'll receive
2. Leave a TODO in the code on that locks with that possible optimization (for now let's leave it like that)
3. Using the metrics data we can decide if we want the optimization or not, and if we do we can probably add a TTL based expiration on the map to ensure it's kept relatively clean


WDYT ?

Reply:
sounds good 👍